src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs

```
@@ -452,17 +452,19 @@ public void ClearOrdering()
                         continue;
                     }
 
-                    if (joinedMapping.Value1 is ColumnExpression && joinedMapping.Value2 is ColumnExpression
-                        || joinedMapping.Value1 is ScalarSubqueryExpression && joinedMapping.Value2 is ScalarSubqueryExpression)
+                    if (joinedMapping.Value1 is SqlExpression innerColumn1
+                        && joinedMapping.Value2 is SqlExpression innerColumn2)
                     {
-                        handleColumnMapping(
-                            joinedMapping.Key,
-                            select1, (SqlExpression)joinedMapping.Value1,
-                            select2, (SqlExpression)joinedMapping.Value2);
+                        // The actual columns may actually be different, but we don't care as long as the type and alias
```

Where are we checking type?
What if 2 columns are same on clrType but one is using conversion to different type in store? Invalid SQL.

---

The problem with checking different types, is that in certain cases it may be fine to do a set operation over different stored types - when the database can implicitly convert between them. For example, the following works fine:

```sql
SELECT CAST(1 AS bigint) UNION SELECT CAST(2 AS smallint);
```

Since the database implicitly up-casts smallint to bigint. Since we have no idea which implicit conversions are supported, it seems more correct to let the user's query through and allow the database to error if it doesn't support it.

What do you think?

---

https://github.com/aspnet/EntityFrameworkCore/issues/15586

---

Thanks, great to see this is already tracked.

Are you OK with leaving it open here for now, until we have the compatibility chart implemented? Since we don't have it at the moment, I'd rather not block valid queries, especially as the consequence for invalid queries is simply a database error (so just a different kind of error).

---

I discussed with @ajcvickers - It is better to throw exception from our code telling what is wrong rather than invalid SQL. So we would make is restrictive and compare store types.
On other note, why the SQL is valid, is it possible to write such linq?
The member's type on both side would be same so at least one of them would have convert to the common type and which would bubble up as cast in SQL anyway which would give same store type.

---

> It is better to throw exception from our code telling what is wrong rather than invalid SQL. So we would make is restrictive and compare store types.

It seems strange for me to block valid queries, just so that for the invalid cases a more informative exception can be raised - but I'll do the change and confirm that the type mappings match.

>  On other note, why the SQL is valid, is it possible to write such linq?
> The member's type on both side would be same so at least one of them would have convert to the common type and which would bubble up as cast in SQL anyway which would give same store type.

I don't think I understand the question...

```sql
SELECT x.TextColumn UNION SELECT x.CaseInsensitiveTextColumn
```

(PostgreSQL really has a case-insensitive text *type* called citext, so the above isn't really contrived).

In LINQ both sides are just strings, but in the database they're different types which are implicitly convertible. This kind of query will no longer work after I add the check that the projection mappings have an identical store type.

---

Just to be clear, I also agree we should be throwing more informative exceptions rather than generating invalid SQL. The point here is that we're also blocking valid queries which would otherwise work.

---

There is easy work-around to add a cast in LINQ query to work-around which does not seem to bad either. Since otherwise given those 2 columns, how are you going to determine which type to read from server.

---

> There is easy work-around to add a cast in LINQ query to work-around which does not seem to bad either.

That's true.

> Since otherwise given those 2 columns, how are you going to determine which type to read from server.

Do you mean in the shaper? Because from a CLR type perspective we're just reading back strings right? I don't see where the problem is (but am probably missing something).

In any case I'll the type mapping checks.

---

Client side is string sure, but shaper also reads values from DataReader. Is it going to read case sensitive string or case insensitive string. It may not matter when there is only clr type but with value converters it becomes altogether a different game. The chart not only tells us about what 2 types are compatible but also need to inform what would be the result type.

---

Good point... ultimately it is best to know exactly which store type is coming out of the set operation, and for that we will need the chart.